### PR TITLE
Update host metrics receiver to use scraper controller

### PR DESCRIPTION
**Pull request by james-bebbington**
Update host metrics receiver to use scraper controller removing around 400 lines of code. There is some "scraper factory" code left over that might be able to be moved into the receiverhelper (or scraper) library. Will think about that for next phase.

Resolves https://github.com/open-telemetry/opentelemetry-collector/issues/918

---

**Sample Output:**

_Note in these examples we see quite a lot of errors scraping process data due to not running the Collector in "Administrative mode" on Windows:_

**Traces (from http://localhost:55679/debug/tracez?zspanname=scraper%2fprocess%2fMetricsScraped&ztype=2&zsubtype=0):**

image

**Metrics (from http://localhost:8888/metrics):**

image

### CODE REVIEW

**Comment by asuresh4** [comment]
@james-bebbington do you think it makes sense to split this PR? Asking since there are changes in `componenterror` and `obsreport` apart from updates to host metrics receiver.

**Comment by james-bebbington** [comment]
> @james-bebbington do you think it makes sense to split this PR? Asking since there are changes in `componenterror` and `obsreport` apart from updates to host metrics receiver.

This PR depends on #1974 (the changes to componenterror, obsreport, etc are all in #1947). If you want you can just review the second commit, but otherwise I would recommend reviewing #1947 first and waiting for that to be merged before reviewing this.

**Comment by james-bebbington** [comment]
This is now ready for review again

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go [hunk 1]
```diff
@@ -135,7 +144,7 @@ func (s *scraper) getProcessMetadata() ([]*processMetadata, error) {
 
 		executable, err := getProcessExecutable(handle)
 		if err != nil {
-			errs = append(errs, fmt.Errorf("error reading process name for pid %v: %w", pid, err))
+			errs = append(errs, consumererror.NewPartialScrapeError(fmt.Errorf("error reading process name for pid %v: %w", pid, err), 1))
```
Comment (jrcamp) [inline receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:147]: Didn't think of this during the previous review but seeing it now a less verbose version might be

```go
var errs consumererror.ScrapeErrors
errs.Add(fmt.Errorf("error reading process name for pid %v: %w", pid, err), 1))
errs.Add(...)
return errs.Combine()
```

Or something. Just an idea.

**Review by jrcamp — APPROVED** [review]

File: receiver/hostmetricsreceiver/internal/scraper.go [hunk 2]
```diff
@@ -75,7 +43,7 @@ type ResourceScraperFactory interface {
 
 	// CreateMetricsScraper creates a resource scraper based on this
 	// config. If the config is not valid, error will be returned instead.
-	CreateMetricsScraper(ctx context.Context, logger *zap.Logger, cfg Config) (ResourceScraper, error)
+	CreateMetricsScraper(ctx context.Context, logger *zap.Logger, cfg Config) (receiverhelper.ResourceMetricsScraper, error)
```
Comment (asuresh4) [inline receiver/hostmetricsreceiver/internal/scraper.go:46]: nit: Consider renaming this method to `CreateResourceMetricsScraper` for clarity?

**Review by asuresh4 — APPROVED** [review]

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go (hunk 1 above)
Comment (james-bebbington) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:147]: It is a bit ugly yea.

I don't know if your suggestion makes that much difference though. Another idea might be to make a `fmt`-like constructor for `PartialScrapeError` so we can do something like `consumererror.FmtPartialScrapeError(1, "error reading process name for pid %v: %w", pid, err)`

WDYT?

In any case I will probably do in a follow up PR as this one is very large already

Comment (james-bebbington) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:147]: Makes sense - we would also want a way to include regular unexpected errors that don't necessarily make sense to be represented as a scrape error.

Moved this discussion to an issue: https://github.com/open-telemetry/opentelemetry-collector/issues/2046

**Pull request merged**